risc0/zkvm/platform/src/heap/bump.rs

```diff
@@ -35,3 +53,58 @@ unsafe impl GlobalAlloc for BumpPointerAlloc {
         self.alloc(layout)
     }
 }
+
+/// Used memory on the heap, in bytes. Note that the bump allocator never frees memory.
+pub fn used() -> usize {
+    // SAFETY: Single threaded, and non-premptive so access is safe.
+    unsafe { HEAP_POS - HEAP_START }
+}
+
+/// Free memory on the heap, in bytes.
+pub fn free() -> usize {
+    // SAFETY: Single threaded, and non-premptive so access is safe. HEAP_POS will always be
+    // less than the start of system memory.
+    GUEST_MAX_MEM - unsafe { HEAP_POS }
+}
+
+pub(crate) unsafe fn alloc_aligned(bytes: usize, align: usize) -> *mut u8 {
+    // SAFETY: Single threaded, and non-premptive so access is safe.
+    let mut heap_pos = unsafe { HEAP_POS };
+
+    // Honor requested alignment if larger than word size.
+    // Note: align is typically a power of two.
+    let align = usize::max(align, WORD_SIZE);
+
+    let offset = heap_pos & (align - 1);
+    if offset != 0 {
+        heap_pos += align - offset;
+    }
+
+    let ptr = heap_pos as *mut u8;
+    heap_pos += bytes;
+
+    // Check to make sure heap doesn't collide with SYSTEM memory.
+    if GUEST_MAX_MEM < heap_pos {
+        const MSG: &[u8] = "Out of memory!".as_bytes();
+        unsafe { sys_panic(MSG.as_ptr(), MSG.len()) };
+    }
+
+    // SAFETY: Single threaded, and non-premptive so modification is safe.
+    unsafe { HEAP_POS = heap_pos };
+    ptr
+}
+
+/// Initialize the bump allocator with the memory allocations defined in the [memory][crate::memory] module.
+///
+/// # Saftey
+///
+/// This function must be called exactly once.
+pub unsafe fn init() {
```

Comment:
Perhaps this should be under an underscored name, like `__init()` and documented that it's called automatically by risc0 boilerplate, given that this `init` call was accidentally called in practice for the embedded alloc?

Reply:
Also curious why this doesn't use something like `OnceCell` as a lazy static for the heap meta-data of each. Assume it might just be for code simplicity and avoiding adding more types to the core API?

Reply:
The __ methods are generally symbols that end up (unmangled) in the ELF. I don't think that helps so much with the risk of someone unwittingly calling this function. I do think the `unsafe` modifier is important here, and communicates the risks to a large extent.

Reply:
> Also curious why this doesn't use something like OnceCell as a lazy static for the heap meta-data of each. Assume it might just be for code simplicity and avoiding adding more types to the core API?

We could do that. I don't really have a strong intuition either way.

Reply:
> We could do that. I don't really have a strong intuition either way.

Just suggesting in an attempt to make the APIs less prone to misuse -- I don't feel strongly about it.

Reply:
OnceCell seems nice

Reply:
OnceCell might also be an unnecessary complication, as it would still require unsafe code to mutate the value. Avoiding init from being called twice could just be as simple as checking if `HEAP_START != 0` in init before modifying to achieve basically the same thing